sale_pricelist_auto_update/models/res_partner.py

```diff
@@ -0,0 +1,130 @@
+            if new_pricelist:
+                if self.property_product_pricelist != new_pricelist:
+                    self.property_product_pricelist = new_pricelist
+            else:
+                raise UserError(
+                    _("Cannot find a pricelist that matches the conditions."))
+        return True
+
+    def _get_customer_ids(self):
+        partners = self.env['res.partner'].sudo().search(
+            [('customer', '=', True),
+             ('active', '=', True)]
+        )
+        return [p.id for p in partners]
+
+    @api.multi
+    def _update_partner_purchase_data(self, amount, date_start, date_end):
+        self.ensure_one()
+        # assumption: only one record should match if any
+        history_recs = self.yearly_sales_history_ids.filtered(
+            lambda x: x.end_date == date_end
+        )
+        if history_recs:
+            if history_recs[0].amt_computed != amount:
+                history_recs[0].write({'amt_computed': amount})
+        else:
+            vals = {
+                'partner_id': self.id,
+                'start_date': date_start,
+                'end_date': date_end,
+                'amt_computed': amount,
+            }
+            self.env['partner.yearly_sales'].sudo().create(vals)
+
+    @api.multi
+    def reset_partner_pricelist(self, date_range, partner_id=False):
+        date_start = date_range.date_start
+        date_end = date_range.date_end
+        if partner_id:
+            partner_ids = tuple([partner_id.id])
+        else:
+            partner_ids = tuple(self._get_customer_ids())
+        ship_pt_recs = self.env['product.template'].search(
+            [('is_shipping_cost', '=', True)])
+        if ship_pt_recs:
+            ship_prod_ids = tuple([pp.id for pp in [pt.product_variant_ids
+                                                    for pt in ship_pt_recs]])
+            params = (ship_prod_ids, partner_ids, date_start, date_end)
+        else:
+            params = (partner_ids, date_start, date_end)
+        sql = """
+            SELECT
+                p.commercial_partner_id AS partner_id,
+                SUM(base_amt) AS amount
+            FROM
+                sale_order_line sol
+            JOIN
+                sale_order so ON sol.order_id = so.id
+            JOIN
+                res_partner p ON so.partner_id = p.id
+            WHERE
+                sol.is_delivery IS FALSE AND
+        """
+        if ship_pt_recs:
+            sql += """
+                sol.product_id NOT IN %s AND
+            """
+        sql += """
+                so.partner_id in %s AND
+                so.state in ('sale', 'done') AND
+                so.date_order_ctx >= %s AND
+                so.date_order_ctx <= %s
+            GROUP BY
+                p.commercial_partner_id
+            """
+        self._cr.execute(sql, params)
+        sales_data = self._cr.dictfetchall()
+        for sales_dict in sales_data:
```

Comment:
If there is not sales order, it will not enter the this loop therefore not update in pricelist etc